### PR TITLE
T-168: asset: .vxs v1 shape-group save format (SHPG, SREF, MODE chunks)

### LINKED ISSUE #665: asset: .vxs v1 shape-group mode (SHPG, SREF chunks) — SDF primitive composition save format
## Area
`engine/asset/`, `engine/prefabs/irreden/voxel/`

## Suggested model
`[opus]` — designs the on-disk shape primitive registry encoding so adding new `ShapeType` values (torus, capsule_knot, custom) doesn't bump the format version. Cross-cutting decision.

## Context

`.vxs` v1 dense mode (#664) handles per-voxel records. This ticket adds the **shape-group mode**: an asset can be saved as a composition of SDF primitive instances instead of a dense voxel grid. The engine already has `C_ShapeDescriptor` (`engine/prefabs/irreden/voxel/components/component_shape_descriptor.hpp:21-37`) and a SHAPES_TO_TRIXEL pipeline (`engine/prefabs/irreden/render/systems/system_shapes_to_trixel.hpp`) — this format persists the inputs to that pipeline.

Shape-group `.vxs` files render directly via SHAPES_TO_TRIXEL (no voxelization required, no voxel pool allocation). Hybrid mode (#668) combines shape primitives with dense voxel overrides.

Designed so that **adding a new `ShapeType` enum value never breaks an old save** — the format writes both numeric id and a string-name table, so an older loader receiving a new save just logs "unknown shape, skipped" and the rest of the asset loads (Extensibility Rule #2).

## Acceptance

- [ ] `.vxs` shape-group chunks defined:
  - `SHPG` chunk: array of primitive instances —
    `{ uint32 shapeTypeId; uint16 version; vec4 params; Color color; uint32 flags; uint8 bone_id; vec3 offset; quat rotation; uint8 csgOp; }`
    where `csgOp` ∈ `{ UNION, SMOOTH_UNION, SUBTRACT, INTERSECT, NONE }`.
  - `SREF` chunk: string name-table for `ShapeType` enum + material registry refs (so a save with a future `ShapeType::TORUS_KNOT` is recoverable on an older build).
- [ ] `.vxs` `MODE` chunk learns the `SHAPES` mode tag (DENSE was #664; HYBRID is #668).
- [ ] `IRAsset::saveVoxelSet` overload accepts `std::span<const C_ShapeDescriptor>` plus optional per-primitive `csgOp`/transform parallel arrays.
- [ ] Loader path: returns a list of `C_ShapeDescriptor`-shaped records the caller spawns as entities.
- [ ] Adding a new `ShapeType` extends the enum in `engine/render/include/irreden/render/ir_render_types.hpp` plus an `sdf<NAME>` GLSL/Metal pair — **no format change required**. Add a "How to add a new SDF primitive" walkthrough to `engine/asset/CLAUDE.md`.
- [ ] Round-trip unit test: build a 5-primitive shape group (mixed sphere/box/capsule), save, load, byte-compare.
- [ ] Edge case: unknown `ShapeType` numeric id (forward-compat) → loader logs "unknown shape ID=N name=…, skipped" and returns the rest of the primitives.

## Critical files

- `engine/asset/include/irreden/asset/voxel_set_format.hpp` (extend)
- `engine/asset/src/voxel_set_format.cpp` (extend)
- `engine/asset/CLAUDE.md` (add SDF-primitive walkthrough)
- `engine/asset/test/test_voxel_set_shapes.cpp` (new)

## Notes

- #664 and this ticket can run in parallel after #663 lands.
- Designed for future CSG-tree extensions; the per-instance `csgOp` byte is sufficient for v1 (flat composition with smooth-union). A tree structure can land in a `SHPT` chunk later without bumping the file version.

## Cross-references

- Parent epic: #604, #213
- Blocked by: #663 (engine/asset/ binary-I/O primitives)
- Blocks: #668 (hybrid mode + sidecar)

### PR DESCRIPTION
**Pull request by jakildev**
## Summary
- `.vxs` v1 SHAPES persistence mode: SDF primitive composition save/load via MODE / SREF / SHPG chunks.
- Forward-compat by name (`ShapeType` registry): future enum values load on older builds as "unknown shape, skipped".
- `C_ShapeDescriptor`-aware prefab adapter (`<irreden/asset/voxel_set_io.hpp>`) writes the same format with parallel offset / rotation / csgOp / boneId arrays.

## Stack context
Stacked on: https://github.com/jakildev/IrredenEngine/pull/678 (T-166 binary I/O primitives).

When #678 merges, GitHub auto-rebases this PR's base to master.

## What's in the diff
- `engine/asset/include/irreden/asset/voxel_set_format.hpp` — container API (MODE / SREF / SHPG primitives, high-level save/load).
- `engine/asset/src/voxel_set_format.cpp` — implementation; canonical `ShapeType` name table.
- `engine/prefabs/irreden/asset/voxel_set_io.hpp` — header-only `IRAsset::saveVoxelSet(span<C_ShapeDescriptor>, ...)` adapter.
- `test/asset/voxel_set_shapes_test.cpp` — 14 GoogleTest cases (round-trip, byte-stable re-serialization, unknown-shape skip, future-build rename, legacy no-SREF save, container errors, future-chunk-tag tolerance).
- `engine/asset/CLAUDE.md`, `engine/prefabs/irreden/asset/CLAUDE.md` — chunk layout doc + "How to add a new SDF primitive" walkthrough + adapter entry.

## Test plan
- [x] `fleet-build --target IrredenEngineAsset` clean.
- [x] `fleet-build --target IrredenEngineTest` clean.
- [x] `fleet-run --timeout 60 IrredenEngineTest` — 533/533 pass (14 new VoxelSetFormat cases included).

## Notes for reviewer
- Format version stays `1` — DENSE (#664) and HYBRID (#668) will reuse the same container under their own chunk tags; MODE chunk dispatches.
- Per-record `recordVersion_` is the additive-field lever (Save Format Extensibility Rule #3); v1 records share one fixed layout.
- SREF lookups go name→current_enum first, id verbatim only when no SREF chunk exists (legacy / hand-rolled minimal saves).
- The `C_ShapeDescriptor` adapter is exercised by inclusion only — its constructor reads `IRRender::getActiveCanvasEntity()`, so a runtime unit test needs a live `RenderManager`. Demos / creations will exercise the save path with a live world.

Closes #665

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by jakildev — COMMENTED** [review]
## Review — T-168: asset: .vxs v1 shape-group save format (SHPG, SREF, MODE chunks)

Stacked on #678 (currently `fleet:approved`). Reviewing the child diff only — upstream is approved separately.

**Verdict:** approve

### Nits

- **`voxel_set_format.cpp:611` — `CsgOp` loaded from disk without range validation.** `static_cast<CsgOp>(csgR.value_)` on an out-of-enumerator byte is valid per C++ scoped-enum rules, but a downstream switch without a `default` case would silently mishandle the value. `ShapeType` has full name-resolution fallback via SREF; `CsgOp` has none. The PR notes say "Tree-shaped CSG would land in a separate SHPT chunk later without bumping the file version" — i.e. `CsgOp` is intentionally not planned for extension. Worth a one-line comment to make the asymmetry explicit rather than letting a future author wonder:
  ```cpp
  // CsgOp is stable — future composition modes land in SHPT chunks,
  // not new CsgOp values, so no name-resolution fallback is needed.
  rec.csgOp_ = static_cast<CsgOp>(csgR.value_);
  ```
  Alternatively, add a bounds guard that defaults to `CsgOp::NONE` with a logged warning for defense-in-depth; either approach is fine.

- **`voxel_set_io.hpp:785-812` — `saveVoxelSet` warns on mismatch but the contract says "stop at the shorter length."** The docstring calls this "asserted with a logged warning," but assertions fire unconditionally in debug builds — logging a warning and continuing is a different contract. If the intent is "programmer error, fail loudly," use `IR_ASSERT`; if the intent is "user-supplied spans may be shorter, degrade gracefully," document it as a soft fallback rather than an assertion. Minor — the current behavior is not wrong, but the docstring and implementation use different vocabulary for the same thing.

### Praise

- **SREF name-resolution for `ShapeType`** — the name-table-based forward-compat mechanism (`disk id → name → current_enum id, skip on miss`) is exactly the right design. Future builds adding enum values get automatic graceful degradation on older loaders without a format version bump.

- **`readShapeGroupChunk` reserve cap** (`countR.value_ < cap ? countR.value_ : cap`) — the corrupted-count protection from T-166 is carried forward correctly; a poisoned `shapeCount` claiming billions of records doesn't pre-allocate.

- **`FutureBuildRenamedSphereResolvesByName` test** — exercises Rule #2 (name wins over numeric id when the disk-side id is renumbered) directly. This is the subtle happy-path case most format tests omit.

- **Byte-stable reserialization test** — save → load → save → byte-compare is exactly the right correctness signal for any format that must survive round-trips through different writer builds.

### Test plan the author should run before merge

- [x] 533/533 `IrredenEngineTest` pass — confirmed in PR body.
- [ ] Optional: add the `CsgOp` comment or bounds guard before merge; no re-review needed for either nit.

Escalation: none. Safe for merge.

Opus recheck not required.

🤖 Reviewed by Claude Sonnet 4.6 (review-pr skill)

**Comment by jakildev** [comment]
Addressed nit: added comment at the CsgOp load site documenting why it lacks SREF-style name-resolution fallback (future composition modes go into a separate SHPT chunk, not new CsgOp values). 14/14 VoxelSetFormat + 9/9 BinaryIO tests still pass.

**Comment by jakildev** [comment]
Merger: cascade-rebased onto updated base PR
#678. Its head ref `claude/T-166-asset-binary-io` was force-
pushed since this PR was last rebased; the child commits
replayed cleanly onto the new upstream tip. Force-pushed
with `--force-with-lease`. CI will re-run.

— fleet merger

**Comment by jakildev** [comment]
Addressed nits:
- Nit 1 (CsgOp comment): already present in commit ebc06393 from a prior pass — no change needed.
- Nit 2 (voxel_set_io.hpp docstring): fixed in 4c2d1d6d. Removed the misleading 'programmer error, asserted with a logged warning' phrasing; replaced with 'logs a warning and degrades gracefully by stopping at the shorter span', which accurately describes the soft-fallback implementation. No behavior change.

fleet:approved remains valid. — sonnet-fleet-1

**Comment by jakildev** [comment]
Merger: base PR #678 merged. Re-targeted this PR from `claude/T-166-asset-binary-io` to `master`. The diff against master may differ from the previous review — reviewer will re-evaluate on next pass.

— fleet merger

**Comment by jakildev** [comment]
Merger: cannot auto-resolve mechanically. The PR has
semantic conflicts with current master that need
judgement-level resolution.

Conflicted files:
- `engine/asset/include/irreden/asset/name_table.hpp` — master: `66b4784b engine/asset: add binary I/O + chunk-table + name-table + JSON sidecar primitives (T-166) (#678)`; PR: `3ebeba23 engine/asset: tighten overflow guard and document nameById lifetime`
- `engine/asset/src/chunk_header.cpp` — master: `66b4784b engine/asset: add binary I/O + chunk-table + name-table + JSON sidecar primitives (T-166) (#678)`; PR: `3ebeba23 engine/asset: tighten overflow guard and document nameById lifetime`

Note: T-166 (commit `10d54eb8`) already landed on master via
PR #678. This branch carries its own copy of that commit
plus a follow-up tightening commit (`3ebeba23`), so a
straight rebase replays the duplicate T-166 and conflicts
with master's reviewed version. Resolution likely involves
dropping `10d54eb8` from the rebase plan and replaying only
the T-168-specific commits + `3ebeba23` on top of the
already-merged T-166.

Labeled `fleet:semantic-conflict` — an opus-worker will
attempt resolution on its next iteration (rebase,
manually resolve, build, push). If the opus-worker also
can't resolve (truly ambiguous, design decision needed),
it will escalate to `human:needs-fix`.

The `fleet:approved` label has been removed if it was set
— the PR no longer represents a reviewed state.

SHA pair: master=fe1217ff1e9703083be1d1568380c797496cbe5d × PR=4c2d1d6d14d928cc53fc2776af4642bc39c06164

— fleet merger

**Review by jakildev — COMMENTED** [review]
## Re-review — T-168: asset: .vxs v1 shape-group save format (SHPG, SREF, MODE chunks)

**Verdict:** approve

### Context

Stacked on #678 (T-166 binary I/O primitives), which has since merged to master. Merger re-targeted this PR's base from `claude/T-166-asset-binary-io` to `master` and set `fleet:stacked-rebase`. Re-evaluating the T-168 delta only — `fleet-pr diff` correctly scopes the diff to T-168 files (voxel_set_format.hpp, voxel_set_format.cpp, voxel_set_io.hpp, voxel_set_shapes_test.cpp, CLAUDE.md updates).

### CONFLICTING state audit

The branch still carries the T-166 commits that are now in master; this is why GitHub shows `mergeable: CONFLICTING`. `fleet-pr diff` scopes to T-168 changes only — no out-of-scope files or oversized churn in the reviewable delta. The branch will need a rebase onto master (dropping the T-166 commits already merged via #678) before the human can click merge; that is a mechanical pre-merge step, not a code-quality finding.

### Nit resolution confirmed

Both nits from the first review were addressed:

- Nit 1 (CsgOp asymmetry comment): present at `engine/asset/src/voxel_set_format.cpp:282-285` — "CsgOp is stable — future composition modes land in a separate SHPT chunk, not new CsgOp values, so no name-resolution fallback needed." Addressed in commit ebc06393.
- Nit 2 (saveVoxelSet docstring): `engine/prefabs/irreden/asset/voxel_set_io.hpp:31-32` now reads "mismatched lengths log a warning and degrade gracefully by stopping at the shorter span, not UB." Accurately matches the soft-fallback implementation. Addressed in commit 4c2d1d6d.

### No new issues in the post-rebase T-168 delta

- ECS: no tick-side getComponent or mid-iteration structural mutations. The prefab adapter lives in `engine/prefabs/irreden/asset/` to keep `engine/asset/` below the component layer — correct.
- Naming: chunk tag constants, record structs, and free functions follow conventions. No anonymous-namespace violations in headers.
- Test coverage: 14 new GoogleTest cases in `voxel_set_shapes_test.cpp`; byte-stable round-trip, unknown-shape skip, future-build name-resolution, container error paths all present. 533/533 pass confirmed in PR body.
- No render, shader, ECS, or audio paths touched — Opus escalation not warranted.

Escalation: none. Safe for merge once the branch is rebased onto master.

Opus recheck not required.

🤖 Reviewed by Claude Sonnet 4.6 (role-sonnet-reviewer)

**Comment by jakildev** [comment]
Resolved semantic conflict: rebased onto current master, dropping the duplicate T-166 commit (10d54eb8) that conflicted with master's reviewed T-166 (PR #678). The follow-up tightening commit (3ebeba23) was dropped automatically by git as "patch contents already upstream" — master's reviewed T-166 already incorporates that overflow guard + nameById lifetime documentation.

Result: the 4 T-168-specific commits (claim, .vxs shape-group implementation, CsgOp asymmetry comment, saveVoxelSet docstring fix) now stack cleanly on master with no replays of T-166 files. Built IrredenEngineTest cleanly and ran 533/533 tests pass on macOS. Reviewer please re-evaluate — the diff against master is now exactly the T-168 delta. — opus-worker

**Review by jakildev — COMMENTED** [review]
## Re-review — T-168: asset: .vxs v1 shape-group save format (SHPG, SREF, MODE chunks)

**Verdict:** approve

### Context

Post-rebase pass. The opus-worker resolved the semantic conflict (dropping the duplicate T-166 commit `10d54eb8` that conflicted with master's reviewed T-166 via PR #678) and pushed the 4 T-168-specific commits cleanly onto master. The PR is now `MERGEABLE`.

`fleet-pr diff 679` scopes to the T-168 delta only — exactly the files reviewed in prior passes:
- `engine/asset/include/irreden/asset/voxel_set_format.hpp` — new
- `engine/asset/src/voxel_set_format.cpp` — new
- `engine/prefabs/irreden/asset/voxel_set_io.hpp` — new
- `engine/asset/CLAUDE.md`, `engine/prefabs/irreden/asset/CLAUDE.md` — doc updates
- `engine/asset/CMakeLists.txt` — adds `voxel_set_format.cpp`
- `test/CMakeLists.txt`, `test/asset/voxel_set_shapes_test.cpp` — test additions

### Post-rebase delta audit

No new files or hunks outside the T-168 scope. The tightening commit (`3ebeba23` — overflow guard + nameById lifetime docs) was correctly dropped by git as "patch contents already upstream" — master's reviewed T-166 (PR #678) already includes those changes.

Both nits from the first review remain resolved:
- **Nit 1 (CsgOp comment):** `voxel_set_format.cpp:611-614` — "CsgOp is stable — future composition modes land in a separate SHPT chunk, not new CsgOp values, so no name-resolution fallback is needed" — present.
- **Nit 2 (saveVoxelSet docstring):** `voxel_set_io.hpp:778` — "mismatched lengths log a warning and degrade gracefully by stopping at the shorter span, not UB" — present and accurate.

### No new issues

- ECS: no tick-side getComponent or mid-iteration structural mutations.
- Naming: all conventions followed; no anonymous-namespace violations in headers.
- Format invariants: MODE/SREF/SHPG chunk ordering not load-bearing (Rule #1 compliant); name-resolution fallback covers Rule #2; per-record `recordVersion_` is the Rule #3 lever.
- Test coverage: 14 GoogleTest cases in `voxel_set_shapes_test.cpp`; all correctness paths covered. 533/533 pass confirmed on macOS by opus-worker.
- No render, shader, ECS, or audio paths touched.

### Test plan

- [x] `IrredenEngineTest` 533/533 pass — confirmed by opus-worker after rebase.
- [x] `IrredenEngineAsset` clean build — confirmed.

Escalation: none. Safe for merge.

Opus recheck not required.

🤖 Reviewed by Claude Sonnet 4.6 (role-sonnet-reviewer)

File: engine/asset/src/voxel_set_format.cpp [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        tag = kModeTagHybrid;
+        break;
+    default:
+        IRE_LOG_ERROR("makeModeChunk: refusing to persist VoxelSetMode::UNKNOWN");
+        return ChunkPayload{kChunkTagMode, {}};
+    }
+    ChunkPayload out;
+    out.tag_ = kChunkTagMode;
+    out.data_.resize(4);
+    std::memcpy(out.data_.data(), tag.data(), 4);
```
Comment (jakildev) [inline engine/asset/src/voxel_set_format.cpp]: Can we assess use of memcpy here and see if cannonizing a resource manager for all things like this . If so file follow up task to address. Use this filewide

**Comment by jakildev** [comment]
Escalated — filed issue #692 for the memcpy canonicalization assessment. Concerns map to: 4 bit-cast sites in `binary_io.cpp` (recommend `std::bit_cast`), 2 chunk-tag pack/unpack sites in `voxel_set_format.cpp` (recommend small `writeTagBytes` / `readTagBytes` helpers), 2 raw-buffer-copy sites in `binary_io.cpp` (leave as primitives).

PR is internally OK to merge if you accept the deferral; re-add `human:needs-fix` to switch to AMEND mode and I'll land the refactor here instead. The follow-up issue is tagged `[sonnet]` — mechanical refactor with clear scope.

— opus-worker

**Comment by jakildev** [comment]
Addressed feedback: memcpy assessment and cleanup.

**Filewide assessment:** two std::memcpy calls existed in the file, both in the MODE chunk read/write path (lines 111 and 125). Every other chunk builder and reader in the file already uses MemoryBinaryWriter/MemoryBinaryReader — the MODE path was the only deviation.

**Fix applied in this commit:** replaced both with the established BinaryWriter::writeBytes / MemoryBinaryReader pattern. Dropped the now-unused <cstring> include. 533/533 tests pass.

**Resource manager verdict:** the existing BinaryWriter/BinaryReader pair in binary_io.hpp is the canonical abstraction for raw byte I/O in this module — no additional resource manager is needed. The two memcpy calls were a consistency gap, not a missing primitive. — sonnet-fleet-2
